src/EOS/Address.cpp

```diff
@@ -0,0 +1,152 @@
+bool Address::extractKeyData(const std::string& string, Address *address) {
+    // verify if the string has one of the valid prefixes
+    Type type;
+    size_t prefixSize;
+    // std::cout << "Checking prefix for " + string + ".\n";
+    // std::cout << "Prefix: " + string.substr(0, Legacy::prefix.size()) + "\n";
+    // std::cout << "Legacy Prefix: " + Legacy::prefix + " (" << sizeof(Legacy::prefix) << "\n";
+    if (string.substr(0, Legacy::prefix.size()) == Legacy::prefix) {
+        type = Type::Legacy;
+        prefixSize = Legacy::prefix.size();
+    } else if (string.substr(0, Modern::R1::fullPubPrefix.size()) == Modern::R1::fullPubPrefix) {
+        type = Type::ModernR1;
+        prefixSize = Modern::R1::fullPubPrefix.size();
+    } else if (string.substr(0, Modern::K1::fullPubPrefix.size()) == Modern::K1::fullPubPrefix) {
+        type = Type::ModernK1;
+        prefixSize = Modern::K1::fullPubPrefix.size();
+    } else {
+        return false;
+    }
+
+    // the binary data will easily fit in a buffer of the size of base58 data
+    size_t size = string.size();
+
+    uint8_t buffer[size];
+    if (!b58tobin(buffer, &size, string.c_str() + prefixSize)) {
+        return false;
+    }
+
+    if (size != KeyDataSize) {
+        return false;
+    }
+
+    // get a pointer past the padding
+    const uint8_t *bufferP = buffer + string.size() - size;
+    if (isValid(bufferP, size, type)) {
+        if (address) {
+            std::memcpy(address->keyData, bufferP, KeyDataSize);
+            address->type = type;
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+/// Initializes a EOS address from a string representation.
+Address::Address(const std::string& string) {
+    if (!Address::extractKeyData(string, this)) {
+        throw std::invalid_argument("Invalid address string!");
+    }
+}
+
+/// Initializes a EOS address from raw bytes
+Address::Address(const Data& data, Type type) : type(type) {
+    if (!isValid(data.data(), data.size(), type)) {
+        throw std::invalid_argument("Invalid byte size!");
+    }
+
+    std::memcpy(keyData, data.data(), KeyDataSize);
+}
+
+/// Initializes a EOS address from a public key.
+Address::Address(const PublicKey& publicKey, Type type) : type(type) {
+    assert(PublicKeyDataSize == TW::PublicKey::secp256k1Size);
+
+    // copy the raw, compressed key data
+    std::memcpy(keyData, 
+                reinterpret_cast<void *>(publicKey.compressed().bytes.data()), 
+                PublicKeyDataSize);
+
+    // append the checksum
+    uint32_t checksum = createChecksum(keyData, type);
+    Data bytes;
+    encode32LE(checksum, bytes);
+
+    for (int i = 0; i < ChecksumSize; ++i) {
+        keyData[PublicKeyDataSize + i] = bytes[i];
+    } 
+}
+
+/// Returns a string representation of the EOS address.
+std::string Address::string() const {
+    size_t b58Size = Base58Size;
+    char b58[b58Size];
+    b58enc(b58, &b58Size, keyData, KeyDataSize);
```

Comment:
we have a c++ implementation in `Base58.h`, it's easier to use

Reply:
At the time of writing this code Base58 was not available. Updating code to reuse the existing base58.